File: Python/core/npc.py
class NPC:
    def __init__(self, npc_id, name, x=0, y=0) :
        self.id = npc_id
        self.name = name
        self.x = x
        self.y = y
        self.behaviour = None
    
    def set_behaviour(self, behaviour) :
        self.behaviour = behaviour

    def move(self, dx, dy) :
        new_x = self.x + dx
        new_y = self.y + dy

        if hasattr(self, 'world') and self.world.is_walkable(new_x, new_y):
            self.x = new_x
            self.y = new_y
            print(f"{self.name} moved to ({self.x}, {self.y})")
            return True
        else:
            print(f"{self.name} cannot move to ({new_x}, {new_y}) — blocked or out of bounds.")
            return False

    def update(self) :
        if self.behaviour:
            self.behaviour.tick(self)
        else:
            print(f"{self.name} is idle at ({self.x}, {self.y})")

    def __repr__(self):
        return f"<NPC {self.name} at ({self.x}, {self.y})>"

File: C_SHARP/NPC.cs
using OpenNPC_CSharp_ver;
using System;

namespace OpenNPC_CSharp_ver
{
	public class NPC
	{
		public int Id { get; }
		public string Name { get; }
		public int X { get; set; }
		public int Y { get; set; }
		public GridWorld World { get; set; }
		public IBehaviour Behaviour { get; set; }

		public NPC(int id, string name, int x, int y)
		{
			Id = id;
			Name = name;
			X = x;
			Y = y;
		}

		public void SetBehaviour(IBehaviour behaviour)
		{
			Behaviour = behaviour;
		}

		public bool Move(int dx, int dy)
		{
			int newX = X + dx;
			int newY = Y + dy;

			if (World.IsWalkable(newX, newY))
			{
				X = newX;
				Y = newY;
				Console.WriteLine($"{Name} moved to ({X}, {Y})");
				return true;
			}
			else
			{
				Console.WriteLine($"{Name} cannot move to ({newX}, {newY}) — blocked or out of bounds.");
				return false;
			}
		}

		public void Update()
		{
			Behaviour?.Tick(this);
		}
	}
}